### ISSUE TITLE
Inconsistent Output of Imported Keras Model with Batch Normalization

### ISSUE THREAD

**Issue by liujxing** [deeplearning4j/deeplearning4j#3167]
#### Issue Description

The output in DL4J of model imported from Keras is different from the output in python, as long as batch normalization layer is used. The discrepancy ranges from 2% to 100% in some extreme cases even with very simple feedforward network.

Json of example network from Keras:

{'class_name': 'Model',
 'config': {'input_layers': [['input_49', 0, 0]],
  'layers': [{'class_name': 'InputLayer',
    'config': {'batch_input_shape': [None, 10],
     'input_dtype': 'float32',
     'name': 'input_49',
     'sparse': False},
    'inbound_nodes': [],
    'name': 'input_49'},
   {'class_name': 'Dense',
    'config': {'W_constraint': None,
     'W_regularizer': None,
     'activation': 'relu',
     'activity_regularizer': None,
     'b_constraint': None,
     'b_regularizer': None,
     'bias': True,
     'init': 'normal',
     'input_dim': 10,
     'name': 'dense_130',
     'output_dim': 30,
     'trainable': True},
    'inbound_nodes': [[['input_49', 0, 0]]],
    'name': 'dense_130'},
   {'class_name': 'BatchNormalization',
    'config': {'axis': -1,
     'beta_regularizer': None,
     'epsilon': 0.001,
     'gamma_regularizer': None,
     'mode': 0,
     'momentum': 0.99,
     'name': 'batchnormalization_57',
     'trainable': True},
    'inbound_nodes': [[['dense_130', 0, 0]]],
    'name': 'batchnormalization_57'},
   {'class_name': 'Dense',
    'config': {'W_constraint': None,
     'W_regularizer': None,
     'activation': 'linear',
     'activity_regularizer': None,
     'b_constraint': None,
     'b_regularizer': None,
     'bias': True,
     'init': 'glorot_uniform',
     'input_dim': 30,
     'name': 'dense_131',
     'output_dim': 1,
     'trainable': True},
    'inbound_nodes': [[['batchnormalization_57', 0, 0]]],
    'name': 'dense_131'}],
  'name': 'model_32',
  'output_layers': [['dense_131', 0, 0]]},
 'keras_version': '1.2.0'}

**Comment by tomthetrainer** [comment]
@turambar We may need more information on this. First however,  I am seeing if @turambar knows if Batch Normalization is an issue or not.

**Comment by huitseeker** [comment]
@crockpotveggies wanna have a look ?

**Comment by turambar** [comment]
@liujxing you using the imported model purely for inference (i.e., you import it and then make predictions without training) or are you doing further training? If it's the latter, I think there may be subtle differences between the way DL4J's and Keras' respective BN layers behave during training.

**Comment by liujxing** [comment]
@turambar Right now only for inference.

**Comment by crockpotveggies** [comment]
My guess here is we're mathematically different from Keras BatchNorm. I'll have to take a look at the underlying mechanics. I may need to poke @AlexDBlack for some help on this. Basically, step-by-step of Keras impl, then look at DL4J impl and check if:

- we have mathematical differences
- params weren't copied appropriately on import
- there's an underlying bug in BatchNorm impl

**Comment by crockpotveggies** [comment]
A note that this commit
https://github.com/deeplearning4j/deeplearning4j/commit/9c2326e4a63ec0d64db0d8a61cd4b518669b5cd0
addresses
a problem in batch norm and LRN. I'll be testing to see if this addresses
this problem.

**Comment by maxpumperla** [comment]
See also #3659 and #3423. @liujxing do you have any insights into how this restricts to batchnorm? i.e. have you seen discrepancies for other layer types and models as well?

I will slowly dive into the root cause of this and update model import in the next weeks.

**Comment by raver119** [comment]
Is this still an issue?

**Comment by maxpumperla** [comment]
haven't seen any indication of this and no feedback, so I'll close this due to inactivity for now. feel free to reopen, though. there are a few other open batchnorm tickets that I'll look into anyway
